userspace/libscap/scap.h

```diff
@@ -1108,6 +1108,14 @@ uint32_t scap_event_decode_params(const scap_evt *e, struct scap_sized_buffer *p
 int32_t scap_event_encode_params(struct scap_sized_buffer event_buf, size_t *event_size, char *error, enum ppm_event_type event_type, uint32_t n, ...);
 int32_t scap_event_encode_params_v(struct scap_sized_buffer event_buf, size_t *event_size, char *error, enum ppm_event_type event_type, uint32_t n, va_list args);
 
+/*!
+  \brief Allocate and create an event from the parameters given as arguments.
+
+  This function behaves exactly like scap_event_encode_params but it will allocate the event and return it by setting the event pointer.
+*/
+int32_t scap_event_generate(scap_evt **event, char *error, enum ppm_event_type event_type, uint32_t n, ...);
```

Comment:
I'm a bit on the fence on this. I agree that it would probably not be used in actual syscall generation code in the hot path since it's slower than managing your own buffer and encoding data when necessary.
On the other hand, this can be useful to implement sinsp tests where we manually encode a series of events, send them to the inspector and then check the final state (without having to record a scap file). In that case we could either replicate the same function there or have this common one do the work.

Reply:
Would it make sense to move this function somewhere inside `userspace/libscap/test/` then? I just feel adding an extra function to the library comes with ramifications that we may not want to deal with if it is only meant to be used for testing.

Reply:
> I just feel adding an extra function to the library comes with ramifications that we may not want to deal with if it is only meant to be used for testing.

Makes sense to me. I'll move it to test code and we can always move it out if we end up using somewhere else. 👍 

> I think instead of using this we can look into [gtest fixtures](https://github.com/google/googletest/blob/main/docs/primer.md#test-fixtures-using-the-same-data-configuration-for-multiple-tests-same-data-multiple-tests) to allocate the event, tests can then format the event itself and call scap_event_decode_params directly, WDYT?

I've looked into fixtures and it looks to me that we cannot allocate the event from there since we don't know how big the event itself will be before we fill it with data, and each test will use different data 🤔

Reply:
> I've looked into fixtures and it looks to me that we cannot allocate the event from there since we don't know how big the event itself will be before we fill it with data, and each test will use different data 🤔

True, but because we are testing, we don't need to worry about being 100% efficient, so you could allocate a buffer large enough to handle any event during setup in the fixture, then encode and decode the event in the test, leaving free'ing of the event to the teardown portion of the fixture. Would that make sense in your opinion?

Another option is allocating a large enough global (to the test) buffer statically and using that without needing to dynamically allocate anything or deal with fixtures.

Reply:
I thought about it a little and I've come up with a possible solution to the teardown annoyance:
* Since we're testing, I want to also exercise the part of the encoding function that computes the length of the event. To do so, I'm running these tests with ASAN enabled (locally for now but would be great to add to CI). This way, if the encoder tries to write past what it says it's the event size I'll get a very clear "Heap-Buffer-Overflow" error. If we only have a giant buffer we would miss such issues. Maybe I'm a bit paranoid but I believe one cannot be too paranoid about memory safety in C and C++
* Since the problem is only about teardown of an event and since gtest is C++ I currently propose to handle that with a smart pointer. This way, if we have `ASSERT`s the resources will be properly cleaned up (and ASAN won't complain about memory leaks :) )

You can take a look at the code, WDYT?

Reply:
I think that should work. One final thing we could consider now that `scap_event_generate` is effectively a C++ function, we could pass a `std::unique_ptr<scap_evt*>&` as the `event` parameter, that way we don't have to concern ourselves with assigning the pointer into a smart pointer in the test itself. Would that make sense?

Reply:
Ah, I missed the conversation where you went the opposite of what I just proposed (moving the function to the library itself).

Personally, I'd very much prefer to avoid random C++ droppings in libscap.

Reply:
I'm fine with using gtest, somewhat less so with a single function out of the entire code base being in C++, even if it happens to live in a test case.